### PR TITLE
WIP: Create new document for rate-limiting issue

### PR DESCRIPTION
**Pull request by MalloZup**
This is a WIP PR for tackling the problem on #3525.

## Status:

Right now is on a initial draft mode, not even worth to be reviewed. 

#### todos:

- add solution proposal .
- add alternatives.

### CODE REVIEW

**Comment by MalloZup** [comment]
just reading about various things. I found also good this source: 
https://cloud.google.com/solutions/rate-limiting-strategies-techniques

**Comment by bwplotka** [comment]
Wow I missed this. Will take a look (:

**Comment by MalloZup** [comment]
@bwplotka  for the moment I wouldn't take look since I will need to work on this more and change drastically. it is more a skeleton right now..

thx for help sofar anyways :cupid:

File: docs/proposals/202006_thanos_grpc_rate_limiting.md [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+owner: bwplotka
+---
+
+### Related Tickets
+
+* https://github.com/thanos-io/thanos/issues/3525 (main)
+
+## Summary and motivation
+
+We want to implement a way to  "rate limiting" concept in thanos, which should prevent to overlaoad CPU and other resource server side which can result with high concurrent request in GRPC.
```
Comment (bwplotka) [inline docs/proposals/202006_thanos_grpc_rate_limiting.md:15]:
```suggestion
We want to implement a way to "rate limiting" concept in Thanos, which should prevent overloading CPU and other resources on the server-side due to high concurrent requests in GRPC.
```

File: docs/proposals/202006_thanos_grpc_rate_limiting.md [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+
+At the point of this writing GRPC has no standard library way to implement a rate-limiter. Allthough there is a interceptor from Grpc ecosystem ( see later)
```
Comment (bwplotka) [inline docs/proposals/202006_thanos_grpc_rate_limiting.md:17]:
```suggestion
At the point of this writing, gRPC has no standard library way to implement a rate-limiter. Although there is an interceptor from gRPC ecosystem ( see later)
```
Let's be consistent it's called `gRPC`

File: docs/proposals/202006_thanos_grpc_rate_limiting.md [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+
+The goal of the  mechanism should be a unified, safe way for rate limiting, both per component but also potentially per tenant.
+
+There are basically 2 ways of solving the problem:
+We should avoid such saturation with a rate-limiting concept or investigate if client queuing is better.
```
Comment (bwplotka) [inline docs/proposals/202006_thanos_grpc_rate_limiting.md:22]: client queuing is also rate limiting concept, no? (:

File: docs/proposals/202006_thanos_grpc_rate_limiting.md [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+* https://github.com/thanos-io/thanos/issues/3525 (main)
+
+## Summary and motivation
+
+We want to implement a way to  "rate limiting" concept in thanos, which should prevent to overlaoad CPU and other resource server side which can result with high concurrent request in GRPC.
+
+At the point of this writing GRPC has no standard library way to implement a rate-limiter. Allthough there is a interceptor from Grpc ecosystem ( see later)
+
+The goal of the  mechanism should be a unified, safe way for rate limiting, both per component but also potentially per tenant.
```
Comment (bwplotka) [inline docs/proposals/202006_thanos_grpc_rate_limiting.md:19]: I would say another goal is to reuse similar rate limit technique for HTTP as for gRPC

**Review by bwplotka — COMMENTED** [review]
Nice work so far! I would add one more goal.

Plus can we choose (you can propose one) a one solution and other put in `Alternatives` section? 

Plus I would recommend pasting this through Grammarly 🤗 There some typos.

**Comment by MalloZup** [comment]
thx @bwplotka for review :sunflower: Since I have experimented a bit today , I do have much more clear vision how to go further with this and also next year implementation..
thx again for review and helps! :sunflower:

**Comment by MalloZup** [comment]
https://github.com/thanos-io/thanos/pull/3636/files contains already a raw implementation :sun_with_face:

File: docs/proposals/202006_thanos_grpc_rate_limiting.md (hunk 4 above)
Comment (MalloZup) [reply docs/proposals/202006_thanos_grpc_rate_limiting.md:19]: yop right

File: docs/proposals/202006_thanos_grpc_rate_limiting.md (hunk 3 above)
Comment (MalloZup) [reply docs/proposals/202006_thanos_grpc_rate_limiting.md:22]: lol right

**Comment by MalloZup** [comment]
you can read the document here.

https://github.com/MalloZup/thanos/blob/RFC-rate-limiting/docs/proposals/202006_thanos_grpc_rate_limiting.md

**Pull request closed without merging**